File: problems/40/solution.java
// Bad solution. Need to re-write

// TreeNode type
// class TreeNode {
//     int val;
//     TreeNode left;
//     TreeNode right;
//     TreeNode(int x) { val = x; left = null; right = null; }
// }
class Solution {
    ArrayList<Integer> result = new ArrayList<Integer>();

    public ArrayList<Integer> inorderTraversal(TreeNode root) {
        // Implement the solution
        _inorder(root);
        return result;
    }


    public void _inorder(TreeNode root) {
        if (root == null)
            return;

        _inorder(root.left);
        result.add(root.val);
        _inorder(root.right);
    }

    public int KthSmallest(TreeNode root, int K) {
        if (root == null)
            return 0;
        _inorder(root);
        return result.get(K - 1);
    }
}

File: problems/40/solution.cpp
// 	 struct TreeNode {
// 	     string val;
// 	     TreeNode* left;
//       TreeNode* right;
//       TreeNode(string x): val(x), left(NULL), right(NULL) {}
//   };
//
//   class Solution {
// 	   public:
// 		  int KthSmallest(TreeNode* tree, int K);
//   };
//
//   Please refrain from using print statements

void inorder(TreeNode* root, vector<int>& order) {
    if (!root)
        return;
    inorder(root->left, order);
    order.push_back(root->val);
    inorder(root->right, order);
}

int Solution::KthSmallest(TreeNode* root, int K) {
    if (!root) {
        return 0;
    }

    vector<int> order;
	inorder(root, order);
    return order[K-1];
}
